src/main/java/hudson/remoting/AbstractByteBufferCommandTransport.java

```diff
@@ -0,0 +1,286 @@
+                        readState = READ_STATE_MORE_HEADER;
+                    }
+                    break;
+                case READ_STATE_MORE_HEADER:
+                    // we have one byte already
+                    readFrameHeader = ChunkHeader.parse(readFrameHeader, data.get());
+                    readFrameRemaining = ChunkHeader.length(readFrameHeader);
+                    readState = READ_STATE_FRAME_BODY;
+                    break;
+                case READ_STATE_FRAME_BODY:
+                    if (data.remaining() < readFrameRemaining) {
+                        readCommandSizes[readCommandIndex] += data.remaining();
+                        readFrameRemaining -= data.remaining();
+                        receiveQueue.put(data);
+                    } else {
+                        readCommandSizes[readCommandIndex] += readFrameRemaining;
+                        int oldLimit = data.limit();
+                        data.limit(data.position() + readFrameRemaining);
+                        receiveQueue.put(data);
+                        data.limit(oldLimit);
+                        readFrameRemaining = 0;
+                        if (ChunkHeader.isLast(readFrameHeader)) {
+                            readState = READ_STATE_COMMAND_READY;
+                        } else {
+                            readState = READ_STATE_NEED_HEADER;
+                        }
+                    }
+                    break;
+                case READ_STATE_COMMAND_READY:
+                    // we have read in a command
+                    readCommandIndex++;
+                    if (readCommandIndex > readCommandSizes.length) {
+                        readCommandSizes = Arrays.copyOf(readCommandSizes, readCommandSizes.length * 2);
+                    }
+                    readCommandSizes[readCommandIndex] = 0;
+                    readState = READ_STATE_NEED_HEADER;
+                    if (receiver != null) {
+                        processCommand();
+                    }
+                    break;
+            }
+        }
+    }
+
+    private void processCommand() throws IOException {
+        try {
+            FastByteBufferQueueInputStream is = new FastByteBufferQueueInputStream(receiveQueue, readCommandSizes[0]);
+            try {
+                ObjectInputStreamEx ois = new ObjectInputStreamEx(is, channel.baseClassLoader, channel.classFilter);
+                receiver.handle(Command.readFrom(channel, ois));
+            } catch (IOException e1) {
+                LOGGER.log(Level.WARNING, "Failed to construct Command", e1);
+            } catch (ClassNotFoundException e11) {
+                LOGGER.log(Level.WARNING, "Failed to construct Command", e11);
+            } finally {
+                int available = is.available();
+                if (available > 0) {
+                    is.skip(available);
+                }
+                IOUtils.closeQuietly(is);
```

Comment:
This is just to ensure that we read whatever we committed to read so that we can stay aligned... not required by the underlying api contract, but nice to have